src/lab11.py

```diff
@@ -428,54 +408,38 @@ def style_length(node, style_name, default_value):
     else:
         return default_value
 
-def paint_clip_path(node, display_list, rect):
-    clip_path = node.style.get("clip-path")
-    if clip_path:
-        percent = parse_clip_path(clip_path)
-        if percent:
-            width = rect.right() - rect.left()
-            height = rect.bottom() - rect.top()
-            reference_val = \
-                math.sqrt(width * width + 
-                    height * height) / math.sqrt(2)
-            radius = reference_val * percent / 100
-            (center_x, center_y) = center_point(rect)
-            display_list.append(CircleMask(
-                center_x, center_y, radius, rect))
-
-def paint_visual_effects(node, display_list, rect):
-    restore_count = 0
-    
+def paint_visual_effects(node, cmds, rect):
+    opacity = float(node.style.get("opacity", "1.0"))
+    paint = skia.Paint(Alphaf=opacity)
+    cmds = [SaveLayer(paint, cmds)]
+
     blend_mode_str = node.style.get("mix-blend-mode")
-    blend_mode = skia.BlendMode.kSrcOver
     if blend_mode_str:
         blend_mode = parse_blend_mode(blend_mode_str)
-
-    opacity = float(node.style.get("opacity", "1.0"))
-    if opacity != 1.0 or blend_mode_str:
-        paint = skia.Paint(Alphaf=opacity, BlendMode=blend_mode)
-        display_list.append(SaveLayer(paint, rect))
-        restore_count = restore_count + 1
-
-    clip_path = node.style.get("clip-path")
-    if clip_path:
-        display_list.append(SaveLayer(skia.Paint(), rect))
-        restore_count = restore_count + 1
-
-    border_radius = node.style.get("border-radius")
-    if border_radius:
-        radius = int(border_radius[:-2])
-        display_list.append(Save(rect))
-        display_list.append(ClipRRect(rect, radius))
-        restore_count = restore_count + 1
-
-    return restore_count
-
-def paint_background(node, display_list, rect):
-    bgcolor = node.style.get("background-color",
-                             "transparent")
-    if bgcolor != "transparent":
-        display_list.append(DrawRect(rect, bgcolor))
+        paint = skia.Paint(BlendMode=blend_mode)
+        cmds = [SaveLayer(paint, cmds)]
+
+    clip_path = node.style.get("clip-path", "")
+    circle_radius = parse_clip_path(clip_path)
+    if circle_radius:
+        width = rect.right() - rect.left()
+        height = rect.bottom() - rect.top()
+        center_x = rect.left() + width / 2
+        center_y = rect.top() + height / 2
+
+        mask_cmds = [DrawCircle(center_x, center_y, circle_radius, skia.ColorWHITE)]
+        paint = skia.Paint(BlendMode=skia.kDstIn)
+        cmds.append(SaveLayer(paint, mask_cmds))
+        cmds = [SaveLayer(skia.Paint(), cmds)]
```

Comment:
The point of line 433 is to isolate the element content from the rest of the page so only the element content is clipped. So for example if we have a paragraph with a background that is clipped into a circle, we want its drawing commands to be:

    SaveLayer([DrawRect, DrawText, DrawText, SaveLayer(DstIn, [DrawCircle])])

I might have this wrong, though—I was attempting to do a pure refactor here of your code and probably got it wrong. Definitely we want to fix it before merging, let's discuss Friday.

Reply:
When you do a dst-in blend, you have to first paint the backdrop, then paint the stuff that has dst-in on it. This is because when you call restore(), that means "immediately blend with the existing stuff", but not stuff that comes after.

Reply:
I think this code does that. It creates a new layer for the element contents, draws the element contents, then creates a new layer for the circle, draws and restores it (thus clipping the element contents) and finally draws the clipped element contents into the page. I've confirmed that if I undo changes to the handling of opacity / blending (discussion above) this generates the same Skia commands as the old code.